tensorflow/core/common_runtime/dml/dml_upload_heap.cc

```diff
@@ -53,7 +63,16 @@ StatusOr<DmlGpuEvent> DmlUploadHeap::BeginUploadToGpu(
   // Map the upload heap and copy the source data into it at the specified
   // offset
   void* upload_heap_data = nullptr;
-  DML_CHECK_SUCCEEDED(chunk->resource->Map(0, nullptr, &upload_heap_data));
+  HRESULT hr = chunk->resource->Map(0, nullptr, &upload_heap_data);
+
+  if (hr == DXGI_ERROR_DEVICE_REMOVED) {
+    HRESULT device_removed_reason = d3d12_device_->GetDeviceRemovedReason();
```

Comment:
This seems kinda fragile... for one, this only checks for one of the many possible device removal reasons (e.g. what if Map() returns DXGI_ERROR_DEVICE_HUNG?) But I wonder what the intended end game is here... since device removals are asynchronous, *any* call into D3D12 (that returns an HRESULT) could potentially return one of the device removed error codes. Are we intending to check every single call into D3D12 for device removal?

Reply:
I assume most of them will happen when executing a command list (given the large number of operations submitted at once), but yeah, we should ideally cover all the cases. Device removal isn't a user error and isn't a programming error from us, so it shouldn't just crash. It can be repro'ed as easily as detaching your surface book.